test/BeanstalkClientConnectionLostTest.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Amp\Beanstalk\Test;
+
+use Amp\Beanstalk\BeanstalkClient;
+use Amp\Delayed;
+use Amp\PHPUnit\AsyncTestCase;
+use Amp\Socket\Server;
+use Amp\Socket\SocketException;
+use function Amp\call;
+use function Amp\Promise\all;
+use function Amp\Promise\timeout;
+use function Amp\Socket\listen;
+
+class BeanstalkClientConnectionLostTest extends AsyncTestCase {
+    const PORT_RANGE_MIN = 50000;
+    const PORT_RANGE_MAX = 65535;
+
+    /** @var Server */
+    private $server;
+
+    public function setUp() {
+        parent::setUp();
+        $this->server = $this->initServer();
+    }
+
+    public function tearDown() {
+        parent::tearDown();
+        $this->server->close();
+    }
+
+    /**
+     * @dataProvider dataProviderReserve
+     * @expectedException \Amp\Beanstalk\ConnectionLostException
+     *
+     * @param $reserveTimeout int|null Seconds
+     * @param $connectionCloseTimeout int Milliseconds
+     * @param $testFailTimeout int Milliseconds
+     * @return \Generator
+     */
+    public function testReserve($reserveTimeout, $connectionCloseTimeout, $testFailTimeout) {
+        $beanstalk = new BeanstalkClient("tcp://". $this->server->getAddress());
+        $connectionClosePromise = call(function ($connectionCloseTimeout) {
+            yield new Delayed($connectionCloseTimeout);
+            $this->server->close();
+        }, $connectionCloseTimeout);
+
+        yield timeout(all([
+            $beanstalk->reserve($reserveTimeout),
+            $connectionClosePromise
+        ]), $testFailTimeout);
```

Comment:
You can use [`setTimeout`](https://github.com/amphp/phpunit-util/blob/99db640d44395cf85289e1bfbde9fbed9f2a45b9/src/AsyncTestCase.php#L170) instead.